add support for CIDR exceptions to CiliumNetworkPolicy

ianvernon opened the pull request:
As part of [Kubernetes 1.8 NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/#the-networkpolicy-resource), ipBlock has been added as a field for ingress / egress rules:

> ipBlock: ipBlock describes a particular CIDR that is allowed to the pods matched by a NetworkPolicySpec’s podSelector. The except entry is a slice of CIDRs that should not be included within an IP Block. Except values will be rejected if they are outside the CIDR range.

Add support for similar 'except' entries in CiliumNetworkPolicy for CIDR. In order to not change the existing API, which just takes a set of CIDRs, a new field has been added for both ingress and egress rules, `fromCIDR` and `toCIDR`, which are comprised of a list of `CIDRule`. A CIDRRule is a CIDR from which traffic is allowed, and a list of exceptions to not allow traffic from, much like the `IPBlock` construct in Kubernetes. Such a policy with this new functionality looks like the following:

```
...
 "egress": [
            {
                "toPorts": [
                    {
                        "ports": [
                            {
                                "port": "80",
                                "protocol": "TCP"
                            }
                        ],
                        "rules": {
                            "http": [
                                {
                                    "path": "/public",
                                    "method": "GET"
                                }
                            ]
                        }
                    }
                ],
                "toCIDR": [
                    "10.0.0.1"
                ],
				"toCIDRSet": [
					{
						"cidr": "10.0.0.0/8",
						"except": [
							"10.96.0.0/12"
						]
					}
				]
            }
        ]
...
```

A new library to do manipulations on CIDRs / IPs was added to calculate the resultant set of CIDRs that allow traffic based on the removal of 'except' from the CIDR to/from which traffic is allowed.

There are still a number of things I want to address before this is merged, but I am putting this out as an RFC. The edges are a bit rough - please confirm that the API design is appropriate.

- [x] Add IPv6 Unit tests
- [x] Add runtime tests
- [x] Policy doc updates (will do once API is approved).

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/k8s: factor out port parsing into function

Factor out parsing of ports into separate helper function in K8s package.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/ip: add IP package

pkg/ip is a new library for parsing and manipulating IPs and IP Networks.
It currently allows for the removing of a set of CIDRs that are subnets of
another CIDR and generating the resultant set of CIDRs from this removal.
It is designed to work for both IPv4 and IPv6 addresses.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: add toCIDRSet and fromCIDRSet

* Add new fields, toCIDRSet and fromCIDRSet to CiliumNetworkPolicy to
Egress and Ingress rules accordingly. These fields are lists of CIDRRule, which
is a structure containing a CIDR to which communication is allowed, and a list
of subnets of that CIDR to which communication is not allowed.
* Transform the allowed CIDR and the excluded subnets of that CIDR into a set
of allowed CIDRs that are then added into the BPF maps.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
tests: import test with CIDR except policy

Add a basic runtime test to import a policy using the new CIDRRule.

Signed-off by: Ian Vernon <ian@cilium.io>

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+// Copyright 2016-2017 Authors of Cilium
```
This is a new file. The copyright should cover 2017 only.

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ip
```
This seems to be a new package, so you need a comment here to describe the whole package.

tgraf replied:
We started doing `doc.go` for each package like Kubernetes does to never drop these comments by accident if we delete a file.

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+	ipv6BitLen = 8 * net.IPv6len
+)
+
+// ByMask is used to sort a list of IP networks by the size of their masks.
```
Document in the comment above (and maybe assert?) that this type implements `sort.Interface`.

ianvernon replied:
Fixed, as well as added assertion.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+)
+
+// ByMask is used to sort a list of IP networks by the size of their masks.
+type ByMask []*net.IPNet
```
Since this is exported, I'd choose a more descriptive name, e.g. `IPNetsByMask`.

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+// contained within the CIDR, the CIDR to remove is ignored. A slice of CIDRs is
+// returned which contains the set of CIDRs provided minus the set of CIDRs which
+// were removed.
+func RemoveCIDRs(allowCIDRs, removeCIDRs []*net.IPNet) ([]*net.IPNet, error) {
```
This function doesn't _remove_ CIDR prefixes, it _excludes_ CIDR prefixes.
And "CIDR" stands for "Classless Inter-Domain Routing", so it's meaningless alone. You should use the term "CIDR prefix" or instead.
So I'd suggest renaming this into:
```
func ExcludeCIDRPrefixes(allowedCIDRPrefixes, excludedCIDRPrefixes []*net.IPNet) ([]*net.IPNet, error) {
...
}
```

tgraf replied:
We use `CIDR` alone everywhere. So does Kubernetes and golang. I don't think it's worth to change it, certainly not worth changing it just here if we use it alone everywhere else.

ianvernon replied:
I did what @tgraf referred to and followed the convention in golang / K8s, so I am leaving this as-is.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+}
+
+// RemoveCIDRs removes the specified CIDRs from another set of CIDRs. If a CIDR to remove is not
+// contained within the CIDR, the CIDR to remove is ignored. A slice of CIDRs is
```
Document that both input slices may be modified when calling this function.

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+			}
+
+			// Only remove CIDR if it is contained in the subnet we are allowing.
+			if allowCIDR.Contains(remove.IP) {
```
To be consistent with the case on line 103 below:
```
if allowCIDR.Contains(removeCIDR.IP.Mask(removeCIDR.Mask)) {
```

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+
+Loop:
+	for j, remove := range removeCIDRs {
+		for i, allowCIDR := range allowCIDRs {
```
Why isn't your `Loop` label within the remove CIDR prefix range loop?
Why do you need to re-iterate over all excluded CIDR prefixes again? That's unnecessary because the removeCIDRs slice is not modified by the inner loop.

ianvernon replied:
You're right! Fixed :)

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+				goto Loop
+			}
+		}
+		removeCIDRs = append(removeCIDRs[:j], removeCIDRs[j+1:]...)
```
That shouldn't be necessary, since you could iterate only once over the excluded prefixes by just moving the `Loop` label within the outer loop.

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+	return allowCIDRs, nil
+}
+
+func getFirstIP(ipNet *net.IPNet) *net.IP {
```
The name of this function is misleading. I would expect it to return the first IP address in the given CIDR prefix. If the prefix length is <32 (IPv4) or <128 (IPv6), then its least-significant bit should be 1 (the host identifier part should not be zero).
What this function does is return the network prefix part of the input prefix, so I'd name this function accordingly: `getNetworkPrefix`.

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+		removeIsIpv4 = true
+	}
+
+	if (removeIsIpv4 && !allowIsIpv4) || (!removeIsIpv4 && allowIsIpv4) {
```
```
if removeIsIpv4 != allowIsIpv4 {
```

ianvernon replied:
Heh, that's just a teeny bit simpler :)

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	if (removeIsIpv4 && !allowIsIpv4) || (!removeIsIpv4 && allowIsIpv4) {
+		return nil, fmt.Errorf("cannot mix different address spaces")
```
`"cannot mix IP addresses of different IP protocol versions"`

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+	// Create CIDR's with mask size of Y+1, Y+2 ... X where Y is the mask length of the CIDR B
+	// from which we are exluding a CIDR A with mask length X.
+	for i := (allowBitLen - allowSize - 1); i >= (allowBitLen - removeSize); i-- {
+		// The mask for each CIDR is simply the ith bit flipped, and then zero'ing
```
CIDR -> CIDR prefix

ianvernon replied:
Fixed. I noticed a typo, 'exluding' which I will fix to be 'excluding', will post with that fix shortly.

rlenglet commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+	// from which we are exluding a CIDR A with mask length X.
+	for i := (allowBitLen - allowSize - 1); i >= (allowBitLen - removeSize); i-- {
+		// The mask for each CIDR is simply the ith bit flipped, and then zero'ing
+		// out all subsequent bits.
```
... subsequent bits (the host identifier part of the prefix).

ianvernon replied:
Fixed.

rlenglet commented on pkg/ip/ip_test.go:
```diff
@@ -0,0 +1,169 @@
+// Copyright 2016-2017 Authors of Cilium
```
2016-2017 -> 2017

ianvernon replied:
Fixed.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -239,6 +265,23 @@ type PortRule struct {
 	Rules *L7Rules `json:"rules,omitempty"`
 }
 
+//CIDRRule is a rule that specifies a CIDR to which outside communication is
```
Add a leading space.

ianvernon replied:
Fixed.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -239,6 +265,23 @@ type PortRule struct {
 	Rules *L7Rules `json:"rules,omitempty"`
 }
 
+//CIDRRule is a rule that specifies a CIDR to which outside communication is
```
to which -> to/from which
Same below.

ianvernon replied:
Fixed.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -239,6 +265,23 @@ type PortRule struct {
+// allowed, along with an optional list of subnets within that CIDR to which
+// outside communication is not allowed.
+type CIDRRule struct {
+
```
Remove this blank line.

ianvernon replied:
Fixed.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -239,6 +265,23 @@ type PortRule struct {
+	// within Cidr.
+	//
+	// +optional
+	ExceptCIDR []CIDR `json:"except,omitempty"`
```
ExceptCIDR -> ExceptCIDRs

ianvernon replied:
Fixed.

rlenglet commented on pkg/policy/api/rule_validation.go:
```diff
@@ -199,3 +213,47 @@ func (cidr CIDR) Validate() error {
+		}
+
+		if len(c.ExceptCIDR) > 0 {
+			return fmt.Errorf("IP was provided, cannot exclude a CIDR from an IP")
```
IP -> IP address
CIDR -> CIDR prefix

ianvernon replied:
I decided to remove this, and ensure that only strings that follow the proper CIDR notation (<IP>/<prefix>) are allowed, as this is what Kubernetes does. If a user wants to supply a single IP, they can provide it as a `/32` or a `/128`.

rlenglet commented on pkg/policy/rule.go:
```diff
@@ -17,7 +17,9 @@ package policy
 
+	"github.com/cilium/cilium/pkg/ip"
 	"github.com/cilium/cilium/pkg/policy/api"
+	"net"
```
Move this import into the other group above.

ianvernon replied:
Fixed.

rlenglet reviewed (commented):
Overall LGTM!

tgraf commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+func getFirstIP(ipNet *net.IPNet) *net.IP {
+	var mask net.IP
+
+	if govalidator.IsIPv6(ipNet.IP.String()) {
```
In the interest of not taking on additional dependencies, can you just use `ipNet.IP.To4() != nil`?

ianvernon replied:
Done, and removed the dependency as well.

tgraf commented on pkg/policy/api/rule.go:
```diff
@@ -143,6 +143,19 @@ type IngressRule struct {
+	// FromCIDRSet is a list of IP blocks which the endpoint subject to the
+	// rule is allowed to receive connections from in addition to FromEndpoints,
+	// along with a list of subnets contained within their corresponding IP block
+	// from which traffic should not be allowed.
```
Needs a comment whether this conflicts with `FromCIDR` or whether both are allowed. If both are allowed, are they related with a logical AND or OR.

ianvernon replied:
Good point. The way I currently have the code structured,`FromCIDR` supersedes `FromCIDRSet`, e.g., if someone specifies `FromCIDR` to have `10.0.0.0/8`, and in the same rule specifies `FromCIDRSet` to have `10.0.0.0/8` except CIDR Prefix `10.96.0.0/12`, then the net result would be to allow all traffic from `10.0.0.0/8`, as the current implementation does not check whether there are overlaps across `FromCIDRSet` and `FromCIDR` in the rule. Is this acceptable?

rlenglet replied:
I think it should be an OR. Adding a prefix into `FromCIDR` or into `FromCIDRSet` with no exclude should be completely equivalent.
An exclude in a `FromCIDRSet` only reduces the allowed IP addresses from that particular `FromCIDRSet`'s prefix, and no other `FromCIDR` or `FromCIDRSet`. Overlaps among `FromCIDR`s and `FromCIDRSet`s is perfectly fine, since it should be an OR semantics.

tgraf replied:
Agreed, OR or disallowed.

ianvernon replied:
Comments describing overlap and scope have been added to each CIDR-related API object.

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -199,3 +213,47 @@ func (cidr CIDR) Validate() error {
+		return err
+	}
+
+	// Don't need to check error as this was done in validation of CIDR.
```
Well, do you check it or not? ;-)

ianvernon replied:
I changed the logic to ensure that only proper CIDR notation is provided (<IP>/<prefix>) instead of accepting single IPs in `CIDRRule` validation.

tgraf commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+		for i, allowCIDR := range allowCIDRs {
+
+			// Don't allow comparison of different address spaces.
+			if !govalidator.IsIPv4(allowCIDR.IP.String()) && govalidator.IsIPv4(remove.IP.String()) ||
```
This can be done in `Validate()` so you don't need to error out here.

ianvernon replied:
This is in its own package, though. If we want to use these functions elsewhere for whatever reason, it is useful to have an error returned.

tgraf commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,219 @@
+	removeSize, _ := removeCIDR.Mask.Size()
+
+	if allowSize >= removeSize {
+		return nil, fmt.Errorf("allow CIDR must be a superset of remove CIDR")
```
This can be done in `Validate()` so you don't need to error out here.

ianvernon replied:
Quoting above comment:
> This is in its own package, though. If we want to use these functions elsewhere for whatever reason, it is useful to have an error returned.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -209,6 +211,27 @@ func mergeL3(ctx *SearchContext, dir string, ipRules []api.CIDR, resMap *L3Polic
+		}
+		resultantAllowedCIDRs, err := ip.RemoveCIDRs([]*net.IPNet{allowNet}, removeSubnets)
+		if err != nil {
+			//TODO - error handling?
```
All of the error conditions can be avoided by moving the code to `Validate()`. Error conditions during policy eval should be avoided at all costs as we have no choice but to lock down the entire endpoint.

ianvernon replied:
Cool, that's what I figured - wanted to make sure of that as part of my initial RFC. Thanks!

ianvernon replied:
Fixed and added comments explaining why we are ignoring errors.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -209,6 +211,27 @@ func mergeL3(ctx *SearchContext, dir string, ipRules []api.CIDR, resMap *L3Polic
+	for _, s := range cidrs {
+		_, allowNet, err := net.ParseCIDR(string(s.Cidr))
+		if err != nil {
+			// TODO what to do with error? it should already be valdiated at this point, can we ignore?
```
Either document that this is impossible or store the parsed `net.IPNet` in the `api.CIDRRule` as private member so you don't have to parse again on evaluation.

ianvernon replied:
Changed the logic here to ignore the error, as all error cases are handled in the validation now. Do you still want me to store the computed `net.IPNet` objects from `api.CIDRRule` in `api.CIDRRule`?

ianvernon commented on pkg/ip/ip.go:
```diff
@@ -0,0 +1,222 @@
+
+// IPNetsByMask is used to sort a list of IP networks by the size of their masks.
+// Implements sort.Interface.
+type IPNetsByMask []*net.IPNet
```
I prefer `IPNetsByMask`. @rlenglet since you suggested this change, what do you think?

rlenglet replied:
Hound got a point. `NetsByMask` sounds OK.

ianvernon replied:
OK, I'll update the type accordingly.

ianvernon replied:
Fixed in latest update.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -209,6 +211,28 @@ func mergeL3(ctx *SearchContext, dir string, ipRules []api.CIDR, resMap *L3Polic
 	return found
 }
 
+func computeResultantCIDRSet(cidrs []api.CIDRRule) []*net.IPNet {
```
Could you return `[]api.CIDR` here instead?

ianvernon replied:
Fixed.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -220,10 +244,31 @@ func (r *rule) resolveL3Policy(ctx *SearchContext, state *traceState, result *L3
+
+		allResultantAllowedCIDRs := computeResultantCIDRSet(r.FromCIDRSet)
+		// Transform back to api.CIDR.
+		for _, s := range allResultantAllowedCIDRs {
```
Returning `[]api.CIDR` would allow to just `allCIDR = append(allCIDR, computeResultantCIDRSet(r.FromCIDRSet)...)`

ianvernon replied:
Fixed.

Commit:
Documentation: update L3 CIDR documentation

Add documentation for new toCIDRSet and fromCIDRSet

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
NEWS.rst: update NEWS

Update news with CIDR API and Documentation changes.

Signed-off by: Ian Vernon <ian@cilium.io>
